quinn/src/streams.rs

```diff
@@ -427,8 +442,25 @@ where
         &mut self,
         cx: &mut Context,
         bufs: &mut [Bytes],
-    ) -> Poll<Result<Option<usize>, ReadError>> {
-        self.poll_read_generic(cx, |conn, stream| conn.inner.read_chunks(stream, bufs))
+    ) -> Poll<Result<usize, ReadError>> {
+        assert!(!bufs.is_empty());
+        self.poll_read_generic(cx, |conn, stream| {
+            let mut chunks = conn.inner.read(stream, true)?;
+            let mut read = 0;
+            loop {
+                let result = chunks.next(usize::MAX);
+                if let Some(chunk) = result.chunk {
+                    bufs[read] = chunk.bytes;
+                    read += 1;
+                    if read == bufs.len() {
+                        return Ok((read, read > 0, result.status, chunks.finalize()));
+                    }
+                }
+                if result.status != ReadStatus::Readable {
+                    return Ok((read, read > 0, result.status, result.should_transmit));
```

Comment:
Are intermediate `.status` and `.should_transmit` results not relevant? I think this would require that the final values always accumulate/contain the previous values.

Reply:
If you look at `Chunks::next()`, `status` will always be `Readable` if `chunk` is not `None`, and `should_transmit` will always be `ShouldTransmit(false)`, by definition. Maybe the definition is wrong, but the handling here seems fine.

Reply:
> status will always be Readable if chunk is not None

Off the cuff, that sounds like maybe we should actually have a `Result<Chunk, Status>`.

Reply:
So maybe the return type should be `(Result<Option<Chunk>, ReadError>, ShouldTransmit)`?

Reply:
Right, similar to the current interface, except with the new iterator layer.

Reply:
Reverted to pretty much the current return types, I think these changes didn't improve things.